resources/blog/posts/2025/2025-11-12-make-phpunit-tests-perfect-in-15-diffs.md

```diff
@@ -0,0 +1,349 @@
+---
+id: 81
+title: "Make PHPUnit tests Perfect in 15 Diffs"
+perex: |
+
+    Rector helps you improve PHP code, upgrade it to latest PHP version, make use of modern features and faster code structures. But did you know it can make your PHPUnit tests faster and easier to read?
+
+    New PHPUnit version have more precise and reliable asserts, but most people don't know about them. They make tests run faster and in case of failure, provide more clear error message you'll understand.
+
+    Rector can help you with that!
+---
+
+There are two main ways to keep your PHPUnit tests up-to-date and in perfect shape without any work.
+
+## First: Use latest PHPUnit features
+
+A year ago we introduced [Composer-version based sets](/blog/introducing-composer-version-based-sets). This feature:
+
+* tells Rector to read your `composer.json`,
+* detect your installed PHPUnit version
+* and automatically pick up the sets that handle the upgrade to new features
+
+<br>
+
+To enable it, just add this line to your `rector.php` config:
+
+```php
+use Rector\Config\RectorConfig;
+
+return RectorConfig::configure()
+    ->withComposerBased(phpunit: true);
+```
+
+
+<br>
+
+Setup once and forget. That's it. No need to list sets everytime new PHPUnit version is out. Rector now automatically applies PHPUnit upgrade sets, everytime you change version in `composer.json`:
+
+```diff
+ {
+     "require-dev": {
+-        "phpunit/phpunit": "^12.5"
++        "phpunit/phpunit": "^13.0"
+     }
+}
+```
+
+<br>
+
+Run composer to install the new PHPUnit version. Then Rector to apply the upgrade:
+
+```bash
+composer update
+vendor/bin/rector
+```
+
+<br>
+
+That's it!
+
+
+<br>
+
+## Second: Use the best PHPUnit asserts and practices
+
+Using new feature is first step to perfection, but there is more. Less code is better, and more precise assertion is better. But PHPUnit has so many different asserts, it's hard to keep up with the best ones to use in that particular situation.
+
+<br>
+
+### 1. Exact Assertions
+
+In prehistorical past, there was only `assertTrue()` and `assertFalse()`. Now, PHPUnit has many more precise assertions you can use:
+
+```diff
+-$this->assertTrue(isset($anything["foo"]), "message");
++$this->assertArrayHasKey("foo", $anything, "message");
+```
+
+<br>
+
+```diff
+-$this->assertTrue(property_exists(new Class, "property"));
++$this->assertClassHasAttribute("property", "Class");
```

Comment:
[this assert method no longer exists](https://github.com/sebastianbergmann/phpunit/issues/4601)